python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -993,6 +843,185 @@ def _raise_on_error(self) -> None:
+        models in a dbt run once they are built. Note that row counts will not be fetched
+        for views, since this requires running the view's SQL query which may be costly.
+
+        Returns:
+            Iterator[Union[Output, AssetMaterialization, AssetObservation, AssetCheckResult]]:
+                A set of corresponding Dagster events for dbt models, with row counts attached,
+                yielded in the order they are emitted by dbt.
+        """
+        return DbtEventIterator(
+            self._fetch_row_counts_inner(),
+            dbt_cli_invocation=self._dbt_cli_invocation,
+        )
+
+    def _fetch_row_counts_inner(
+        self,
+    ) -> Iterator[
+        Union[
+            Output,
+            AssetMaterialization,
+            AssetObservation,
+            AssetCheckResult,
+        ]
+    ]:
+        logger.info("Row counts will be fetched for non-view models once they are materialized.")
+
+        # We keep a list of emitted Dagster events and pending futures which augment
+        # emitted events with additional metadata. This ensures we can yield events in the order
+        # they are emitted by dbt.
+        output_events_and_futures: List[Union[Future, DbtDagsterEventType]] = []
+
+        # Point at project directory to ensure dbt adapters run correctly
+        with pushd(str(self._dbt_cli_invocation.project_dir)), ThreadPoolExecutor(
+            max_workers=STREAM_EVENTS_THREADPOOL_SIZE
+        ) as executor:
+            self._dbt_run_thread = executor.submit(
+                self._stream_dbt_events_and_enqueue_postprocessing,
+                output_events_and_futures,
+                executor,
+            )
+
+            # Step through the list of output events and futures, yielding them in order
+            # once they are ready to be emitted
+            event_to_emit_idx = 0
+            while True:
+                all_work_complete = get_future_completion_state_or_err(
+                    [self._dbt_run_thread, *output_events_and_futures]
+                )
+                if all_work_complete and event_to_emit_idx >= len(output_events_and_futures):
+                    break
+
+                if event_to_emit_idx < len(output_events_and_futures):
+                    event_to_emit = output_events_and_futures[event_to_emit_idx]
+
+                    if isinstance(event_to_emit, Future):
+                        # If the next event to emit is a Future (waiting on postprocessing),
+                        # we need to wait for it to complete before yielding the event.
+                        try:
+                            event = event_to_emit.result(timeout=0.1)
+                            yield event
+                            event_to_emit_idx += 1
+                        except:
+                            pass
+                    else:
+                        yield event_to_emit
+                        event_to_emit_idx += 1
```

Comment:
Let's refactor so that there's only one branch that does a `yield`

```suggestion
                    # If the next event to emit is a Future (waiting on postprocessing),
                    # we need to wait for it to complete before yielding the event.
                    try:
                        event = event_to_emit.result(timeout=0.1) if isinstance(event_to_emit, Future) else event_to_emit
                        yield event
                        event_to_emit_idx += 1
                    except:
                        pass
```